### ISSUE TITLE
inline parameter not honored when function is called in an overload implementation

### ISSUE THREAD

**Issue by ARF1** [numba/numba#5142]
Numba fails to honour the `inline` parameter in the following situation with 0.48dev0:

```python
import os

import numba
from numba import ir, njit, types
from numba.extending import overload, register_jitable

os.environ["NUMBA_DEBUG_PRINT_AFTER"] = "ir_legalization"


def bar(x):
    """A function stub to overload"""
    pass


def generate_impl(x):
    @register_jitable(inline="always")
    def impl(x):
        return x[0]

    return impl


@overload(bar)
def ol_bar_tuple(x):
    return generate_impl(x)


@njit
def foo():
    a = bar((1, 2, 3))
    return a


foo()
```

The problem only occurs if `impl` is defined within `generate_impl`. If the definition of `impl` is pulled out, it works just fine.

`NUMBA_DEBUG_PRINT_AFTER` output:
```

-------------------------__main__.generate_impl.<locals>.impl: nopython: AFTER ir_legalization--------------------------
label 0:
    x = arg(0, name=x)                       ['x']
    $const4.1 = const(int, 0)                ['$const4.1']
    $6binary_subscr.2 = static_getitem(value=x, index=0, index_var=$const4.1) ['$6binary_subscr.2', '$const4.1', 'x']
    del x                                    []
    del $const4.1                            []
    $8return_value.3 = cast(value=$6binary_subscr.2) ['$6binary_subscr.2', '$8return_value.3']
    del $6binary_subscr.2                    []
    return $8return_value.3                  ['$8return_value.3']

-------------------------------------__main__.foo: nopython: AFTER ir_legalization--------------------------------------
label 0:
    $2load_global.0 = global(bar: <function bar at 0x0000011C57036CA8>) ['$2load_global.0']
    $const_1 = const(int, 1)                 ['$const_1']
    $const_2 = const(int, 2)                 ['$const_2']
    $const_3 = const(int, 3)                 ['$const_3']
    $const4.1 = build_tuple(items=[Var($const_1, <ipython-input-1-885a339d68c5>:30), Var($const_2, <ipython-input-1-885a339d68c5>:30), Var($const_3, <ipython-input-1-885a339d68c5>:30)]) ['$const4.1', '$const_1', '$const_2', '$const_3']
    del $const_3                             []
    del $const_2                             []
    del $const_1                             []
    $6call_function.2 = call $2load_global.0($const4.1, func=$2load_global.0, args=[Var($const4.1, <ipython-input-1-885a339d68c5>:30)], kws=(), vararg=None) ['$2load_global.0', '$6call_function.2', '$const4.1']
    del $const4.1                            []
    del $2load_global.0                      []
    a = $6call_function.2                    ['$6call_function.2', 'a']
    del $6call_function.2                    []
    $12return_value.4 = cast(value=a)        ['$12return_value.4', 'a']
    del a                                    []
    return $12return_value.4                 ['$12return_value.4']
```

**Comment by esc** [comment]
@ARF1 thanks for reporting this. Can you please specify further what you mean by " If the definition of impl is pulled out"? 

I tried running the following:

```
import os

import numba
from numba import ir, njit, types
from numba.extending import overload, register_jitable

os.environ["NUMBA_DEBUG_PRINT_AFTER"] = "ir_legalization"


def bar(x):
    """A function stub to overload"""
    pass


@register_jitable(inline="always")
def impl(x):
    return x[0]


def generate_impl(x):

    return impl


@overload(bar)
def ol_bar_tuple(x):
    return generate_impl(x)


@njit
def foo():
    a = bar((1, 2, 3))
    return a


foo()
```

And I got:

```
-------------------------------------__main__.impl: nopython: AFTER ir_legalization-------------------------------------
label 0:
    x = arg(0, name=x)                       ['x']
    $const4.1 = const(int, 0)                ['$const4.1']
    $6binary_subscr.2 = static_getitem(value=x, index=0, index_var=$const4.1) ['$6binary_subscr.2', '$const4.1', 'x']
    del x                                    []
    del $const4.1                            []
    $8return_value.3 = cast(value=$6binary_subscr.2) ['$6binary_subscr.2', '$8return_value.3']
    del $6binary_subscr.2                    []
    return $8return_value.3                  ['$8return_value.3']

-------------------------------------__main__.foo: nopython: AFTER ir_legalization--------------------------------------
label 0:
    $2load_global.0 = global(bar: <function bar at 0x10b3cb200>) ['$2load_global.0']
    $const_1 = const(int, 1)                 ['$const_1']
    $const_2 = const(int, 2)                 ['$const_2']
    $const_3 = const(int, 3)                 ['$const_3']
    $const4.1 = build_tuple(items=[Var($const_1, foo103.py:32), Var($const_2, foo103.py:32), Var($const_3, foo103.py:32)]) ['$const4.1', '$const_1', '$const_2', '$const_3']
    del $const_3                             []
    del $const_2                             []
    del $const_1                             []
    $6call_function.2 = call $2load_global.0($const4.1, func=$2load_global.0, args=[Var($const4.1, foo103.py:32)], kws=(), vararg=None) ['$2load_global.0', '$6call_function.2', '$const4.1']
    del $const4.1                            []
    del $2load_global.0                      []
    a = $6call_function.2                    ['$6call_function.2', 'a']
    del $6call_function.2                    []
    $12return_value.4 = cast(value=a)        ['$12return_value.4', 'a']
    del a                                    []
    return $12return_value.4                 ['$12return_value.4']
```

**Comment by ARF1** [comment]
@esc Sorry in attempting to make a minimal reproducer I have simplified away the problem. I thought I had it because a `call ...` was still present but that was from the overload.

So let me try again:
```python
import os

import numba
from numba.extending import overload, register_jitable
from numba import njit, types


os.environ["NUMBA_DEBUG_PRINT_AFTER"] = "ir_legalization"


def bar(x):
    """A function stub to overload"""
    pass


@register_jitable(inline='always')
def impl(x):
    return x[0]


@overload(bar)
def ol_bar_tuple(x):
    def impl_wrapper(x):
        return impl(x)
    return impl_wrapper


@njit
def foo():

    # This will resolve via `ol_bar_tuple` as the argument is a types.UniTuple
    # instance. It will always be inlined as specified in the decorator for this
    # overload.
    a = bar((1, 2, 3))

    return a


foo()
```

Which gives:
```
-------------------------------------__main__.impl: nopython: AFTER ir_legalization-------------------------------------
label 0:
    x = arg(0, name=x)                       ['x']
    $const4.1 = const(int, 0)                ['$const4.1']
    $6binary_subscr.2 = static_getitem(value=x, index=0, index_var=$const4.1) ['$6binary_subscr.2', '$const4.1', 'x']
    del x                                    []
    del $const4.1                            []
    $8return_value.3 = cast(value=$6binary_subscr.2) ['$6binary_subscr.2', '$8return_value.3']
    del $6binary_subscr.2                    []
    return $8return_value.3                  ['$8return_value.3']

----------------------__main__.ol_bar_tuple.<locals>.impl_wrapper: nopython: AFTER ir_legalization----------------------
label 0:
    x = arg(0, name=x)                       ['x']
    $2load_global.0 = global(impl: <function impl at 0x0000027B83AD6678>) ['$2load_global.0']
    $6call_function.2 = call $2load_global.0(x, func=$2load_global.0, args=[Var(x, <ipython-input-1-dce8ba3e5ac3>:26)], kws=(), vararg=None) ['$2load_global.0', '$6call_function.2', 'x']
    del x                                    []
    del $2load_global.0                      []
    $8return_value.3 = cast(value=$6call_function.2) ['$6call_function.2', '$8return_value.3']
    del $6call_function.2                    []
    return $8return_value.3                  ['$8return_value.3']

-------------------------------------__main__.foo: nopython: AFTER ir_legalization--------------------------------------
label 0:
    $2load_global.0 = global(bar: <function bar at 0x0000027B84A685E8>) ['$2load_global.0']
    $const_1 = const(int, 1)                 ['$const_1']
    $const_2 = const(int, 2)                 ['$const_2']
    $const_3 = const(int, 3)                 ['$const_3']
    $const4.1 = build_tuple(items=[Var($const_1, <ipython-input-1-dce8ba3e5ac3>:36), Var($const_2, <ipython-input-1-dce8ba3e5ac3>:36), Var($const_3, <ipython-input-1-dce8ba3e5ac3>:36)]) ['$const4.1', '$const_1', '$const_2', '$const_3']
    del $const_3                             []
    del $const_2                             []
    del $const_1                             []
    $6call_function.2 = call $2load_global.0($const4.1, func=$2load_global.0, args=[Var($const4.1, <ipython-input-1-dce8ba3e5ac3>:36)], kws=(), vararg=None) ['$2load_global.0', '$6call_function.2', '$const4.1']
    del $const4.1                            []
    del $2load_global.0                      []
    a = $6call_function.2                    ['$6call_function.2', 'a']
    del $6call_function.2                    []
    $12return_value.4 = cast(value=a)        ['$12return_value.4', 'a']
    del a                                    []
    return $12return_value.4                 ['$12return_value.4']
```

Now in `__main__.ol_bar_tuple.<locals>.impl_wrapper` there is a `call` to `impl` even though it had the `inline='always'`.

I am also struggling to find a sensible title for this issue. Please change it if you can think of something useful.

**Comment by esc** [comment]
@ARF1: thanks for reporting back to clarify the issue. This is something that we can use to debug further.

**Comment by esc** [comment]
@ARF1 a side-questions: why do you want to inline the function in the first place? Are you expecting to see a performance benefit.

**Comment by esc** [comment]
@ARF1 i looked into this some more and found out, that the `inline` parameter is mainly to allow type inference across function boundaries. For example:

```
from numba import njit


@njit()
def bar(a):
    a.append(10)


@njit
def foo():
    z = []
    bar(z)


foo()
```

If you run this code, it won't work because `z` ends up being imprecise (or unrefined). If you now add `inline='always'` to `bar`, the code of `bar` with the `append` call will be inlined and the code can now be compiled and run.

**Comment by ARF1** [comment]
@esc Many thanks for looking into this.

In short: My fundamental reason was that I wanted to diagnose/fix performance issues. Switching on inlining seemed like a quick thing to do just to exclude it as a root cause.

Expanding on the above:
1. My first step was to `inspect_cfg()`. This gave me almost nothing: It was only a stub with a call to a function that was produced by a generator and which was not easily available for inspection.
2. I thought: no problem, inlining will take care of 2 things at once. It will exclude inlining-issues as a reason for my performance issues and will let me allow to inspect more easily.
3. I had the hope that inlining in IR would help the llvm optimizer. Though I do not have any basis for that. Just thought I would give it a try...
4. The stub-function with call to second function implementing the content is also found in the assembly. This seems slightly wasteful but admittedly it is probably not an issue in the grand scheme of things...

**Comment by esc** [comment]
@ARF1 thanks for following up on this. I looked into this further and there are essentially two types of inlining, the Numba_IR inlining which helps to perform type inference across function boundaries and the LLVM inlining which helps to improve performance. So in short: the type of inlining you attempted is unlikely to improve performance. In any case, it does appear though that you have discovered a bug regardless.

**Comment by esc** [comment]
@stuartarchibald what is your take on this? Do you think that `impl` should have been inlined into the `impl_wrapper`?

**Comment by esc** [comment]
Looking at: http://numba.pydata.org/numba-doc/latest/developer/inlining.html?highlight=inline it appears as though `register_jitable` doesn't even support `inline`. Although that may just be the documentation that is out of date..

**Comment by esc** [comment]
Right, so looking at this further: https://github.com/numba/numba/blob/master/numba/extending.py#L121

`register_jitable` is just a wrapper for `overload` so it probably *should* work.

**Comment by esc** [comment]
@ARF1 I added some clarification to the docs, I would appreciate a review from you, if you have time:

https://github.com/numba/numba/pull/5155

**Comment by stuartarchibald** [comment]
Whilst support for this is not stated, it probably ought to "just work". I think that this is probably just a bug, here:
https://github.com/numba/numba/blob/d7953a18dbf5ea231dc16e967ce8e9b754578ea6/numba/extending.py#L138-L141
`kwargs` are forwarded as `jit_options`, but `@overload` has an explicit `inline` kwarg:
https://github.com/numba/numba/blob/d7953a18dbf5ea231dc16e967ce8e9b754578ea6/numba/extending.py#L56
hence the flag ends up being ignored. Would expect this patch to fix it:
```
--- a/numba/extending.py
+++ b/numba/extending.py
@@ -137,7 +137,8 @@ def register_jitable(*args, **kwargs):
     """
     def wrap(fn):
         # It is just a wrapper for @overload
-        @overload(fn, jit_options=kwargs, strict=False)
+        inline = kwargs.pop('inline', 'never')
+        @overload(fn, jit_options=kwargs, inline=inline, strict=False)
         def ov_wrap(*args, **kwargs):
             return fn
         return fn
```

**Comment by esc** [comment]
@stuartarchibald thanks for the suggestion, will try that.

**Comment by esc** [comment]
@stuartarchibald that seem to fix it for the case above, I now get the following Numba IR

```
-------------------------------------__main__.impl: nopython: AFTER ir_legalization-------------------------------------
label 0:
    x = arg(0, name=x)                       ['x']
    $const4.1 = const(int, 0)                ['$const4.1']
    $6binary_subscr.2 = static_getitem(value=x, index=0, index_var=$const4.1) ['$6binary_subscr.2', '$const4.1', 'x']
    del x                                    []
    del $const4.1                            []
    $8return_value.3 = cast(value=$6binary_subscr.2) ['$6binary_subscr.2', '$8return_value.3']
    del $6binary_subscr.2                    []
    return $8return_value.3                  ['$8return_value.3']

----------------------__main__.ol_bar_tuple.<locals>.impl_wrapper: nopython: AFTER ir_legalization----------------------
label 0:
    x = arg(0, name=x)                       ['x']
    x.0 = x                                  ['x', 'x.0']
    del x                                    []
    $const4.1.2 = const(int, 0)              ['$const4.1.2']
    $6binary_subscr.2.3 = getitem(value=x.0, index=$const4.1.2) ['$6binary_subscr.2.3', '$const4.1.2', 'x.0']
    del x.0                                  []
    del $const4.1.2                          []
    $8return_value.3.4 = $6binary_subscr.2.3 ['$6binary_subscr.2.3', '$8return_value.3.4']
    del $6binary_subscr.2.3                  []
    $6call_function.2 = $8return_value.3.4   ['$6call_function.2', '$8return_value.3.4']
    del $8return_value.3.4                   []
    $8return_value.3 = cast(value=$6call_function.2) ['$6call_function.2', '$8return_value.3']
    del $6call_function.2                    []
    return $8return_value.3                  ['$8return_value.3']

-------------------------------------__main__.foo: nopython: AFTER ir_legalization--------------------------------------
label 0:
    $2load_global.0 = global(bar: <function bar at 0x10407e200>) ['$2load_global.0']
    $const4.1 = const(int, 1)                ['$const4.1']
    $const6.2 = const(int, 2)                ['$const6.2']
    $const8.3 = const(int, 3)                ['$const8.3']
    $10build_list.4 = build_list(items=[Var($const4.1, foo105.py:30), Var($const6.2, foo105.py:30), Var($const8.3, foo105.py:30)]) ['$10build_list.4', '$const4.1', '$const6.2', '$const8.3']
    del $const8.3                            []
    del $const6.2                            []
    del $const4.1                            []
    $12call_function.5 = call $2load_global.0($10build_list.4, func=$2load_global.0, args=[Var($10build_list.4, foo105.py:30)], kws=(), vararg=None) ['$10build_list.4', '$12call_function.5', '$2load_global.0']
    del $2load_global.0                      []
    del $10build_list.4                      []
    a = $12call_function.5                   ['$12call_function.5', 'a']
    del $12call_function.5                   []
    $18return_value.7 = cast(value=a)        ['$18return_value.7', 'a']
    del a                                    []
    return $18return_value.7                 ['$18return_value.7']
```

I also tried the following, but it did not go so smoothly:

```
from numba import njit
from numba.extending import register_jitable


@register_jitable(inline='always')
def bar(a):
    a.append(10)


@njit
def foo():
    z = []
    bar(z)
    return z


print(foo())
```

**Comment by ARF1** [comment]
@stuartarchibald 
> Would expect this patch to fix it:
> 
> ```
> --- a/numba/extending.py
> +++ b/numba/extending.py
> @@ -137,7 +137,8 @@ def register_jitable(*args, **kwargs):
>      """
>      def wrap(fn):
>          # It is just a wrapper for @overload
> -        @overload(fn, jit_options=kwargs, strict=False)
> +        inline = kwargs.pop('inline', 'never')
> +        @overload(fn, jit_options=kwargs, inline=inline, strict=False)
>          def ov_wrap(*args, **kwargs):
>              return fn
>          return fn
> ```

Thanks, that works. I was looking at this section of code but failed to spot the problem. 

It took me some time confirm because I stumbled across another issue with inlining:

```python
def generator():
    closure_var = 123

    @register_jitable(inline='always')
    def test_impl(x):
        assert x == closure_var

    return test_impl
```

This results in a `TypingError` on the assertion: `UnsupportedError: Unsupported constraint encountered: raise $10load_global.0`.

Do you want me to make a reproducer and bug issue for this? On reflection that seems to be a fairly understandable error and probably something that is not going to be supported anyway. Maybe just a note in the documentation?

**Comment by ARF1** [comment]
> @ARF1 I added some clarification to the docs, I would appreciate a review from you, if you have time:

@esc Looks good to me. 

I probably would have still tried inlining just to solve my `inspect_cfg()` problem. But clearly I would have been abusing the tool...

**Comment by esc** [comment]
If we are certain this is a good fix I can prep a PR for it. The question about how to test this remains, I would probably try to compile it a very simple variant and then check the Numba IR for successful inlining.
